### PR TITLE
fix(tasks): enforce input request key uniqueness over a task's lifetime

### PR DESCRIPTION
**Pull request by joshrotenberg**
Point 5 from #1246, filed separately as a focused correctness change.

## The rule

SEP-2663 is explicit:

> Each request key in `inputRequests` **MUST** be unique over the lifetime of a single task. A server **MUST NOT** reuse a key for a subsequent server-to-client request after a response for that key has been delivered, and **MUST NOT** use the same key to refer to two distinct requests over a task's lifetime. This guarantees that `inputResponses` keyed by the same identifier always refer to the request the client expects, eliminates ambiguity for clients deduplicating across polls, and lets servers ignore `inputResponses` for unknown or already-satisfied requests.

## What we did instead

`MemoryTaskStore::require_input` cleared a key's history on reissue:

```rust
// A re-issued key is a fresh question, whatever its prior fate.
for key in requests.keys() {
    task.answered_input_keys.remove(key);
    task.superseded_input_keys.remove(key);
}
```

This was deliberate, and a test named `reissued_key_becomes_a_fresh_question` asserted it. It is still a conformance bug: our own `apply_input_responses` relies on the spec guarantee to ignore already-satisfied keys, so permitting reuse makes the two halves disagree about what a key means.

## The change

A key is spent once issued, whether it was answered or superseded. Reissuing one returns an error naming it, rather than silently starting a second question under an identifier the client may already have answered.

A refused park previously only logged a warning, which left the task `working` with nothing outstanding, a state no `tasks/update` can move. It now fails the task with the reason, matching how an empty request set is already handled.

Uniqueness is scoped to the task. Per the SEP, keys from a preceding MRTR phase are a separate namespace and do not constrain a task's keys, so nothing here reaches across that boundary.

## Tests

`reissued_key_becomes_a_fresh_question` encoded the old behaviour and is replaced by three:

- an answered key cannot be reissued, and the message names it
- a superseded key cannot be reissued either, since the SEP forbids one key naming two distinct requests regardless of whether the first was answered
- distinct keys across rounds still work, including asking again under a new name after an answer

Plus an end-to-end router test: a handler that always asks for the same key parks once, gets answered, asks again on resume, and the task ends `failed` with the key named rather than stranded in `working`.

Full suite, clippy, `-Dmissing-docs`, and the 17-entry feature-combinations matrix are green.

### CODE REVIEW

File: crates/tower-mcp/src/async_task.rs [hunk 1]
```diff
@@ -724,16 +734,38 @@ impl TaskStore for MemoryTaskStore {
...
+        //
+        // That guarantee is what lets a client deduplicate across polls and
+        // lets a server ignore responses for already-satisfied requests, so
+        // reissuing a key is rejected rather than quietly accepted (#1246).
+        let reused: Vec<String> = requests
+            .keys()
+            .filter(|key| {
+                task.answered_input_keys.contains(*key)
+                    || task.superseded_input_keys.contains(*key)
+                    || task.input_requests.contains_key(*key)
```
Comment (joshrotenberg) [inline crates/tower-mcp/src/async_task.rs]: **[P1] This also rejects a key that is still outstanding, but that is not necessarily key reuse.** `require_input` replaces the complete outstanding snapshot, and SEP-2663 forbids using a key for a subsequent/distinct request; the same request may remain outstanding across snapshots. For example, `{approval} -> {approval, region}` currently returns `Err`, after which the router fails the task, even though `approval` was merely carried forward. Please distinguish retained current entries from answered/superseded keys (and reject a current key only if it denotes a changed request), then add a transition test for retaining A while adding B.

File: crates/tower-mcp/src/async_task.rs [hunk 2]
```diff
@@ -467,6 +467,16 @@ pub trait TaskStore: Send + Sync + 'static {
     ///
     /// Returns `Ok(false)` if the task is unknown, expired, or already
     /// terminal.
+    /// # Key uniqueness
+    ///
+    /// SEP-2663 requires every request key to be unique over a single task's
+    /// lifetime: a key is spent once it has been issued, whether it was
+    /// answered or superseded, and must never name a second request. An
+    /// implementation is expected to return an error rather than reissue one,
```
Comment (joshrotenberg) [inline crates/tower-mcp/src/async_task.rs]: The preceding contract at lines 464-466 still says the exact opposite: a reissued key is a fresh question even when previously answered. Please rewrite that paragraph and make the custom-store requirement explicit, including that checking/reserving the issued key and replacing the snapshot must be atomic. Existing external `TaskStore` implementations continue to compile with the old semantics, so this behavioral migration should also be called out in release notes (or made compile-visible if that is the intended compatibility boundary).

File: crates/tower-mcp/src/async_task.rs [hunk 3]
```diff
@@ -724,16 +734,38 @@ impl TaskStore for MemoryTaskStore {
...
+            .keys()
+            .filter(|key| {
+                task.answered_input_keys.contains(*key)
+                    || task.superseded_input_keys.contains(*key)
+                    || task.input_requests.contains_key(*key)
+            })
+            .cloned()
+            .collect();
+        if !reused.is_empty() {
+            return Err(TaskStoreError::Backend(format!(
```
Comment (joshrotenberg) [inline crates/tower-mcp/src/async_task.rs]: `TaskStoreError::Backend` is documented as a storage/connection failure—possibly transient—and the enum docs still say `MemoryTaskStore` never returns errors. Key reuse is a deterministic state-machine violation, not a backend failure. Could this get a semantic variant such as `InvalidTransition`/`InputKeyReused` (and corresponding router handling), or otherwise update the error contract? Conflating these prevents callers from distinguishing retryable infrastructure failure from invalid task behavior.

**Review by joshrotenberg — COMMENTED** [review]
The SEP-2663 lifetime rule and the separate MRTR namespace both check out, and CI plus the focused task tests are green. I found one functional snapshot edge case and two public-contract/error-taxonomy issues below. Minor cleanup: the point-1 stray-update doc block at `router.rs:6541-6544` is now attached to the new point-5 test; it should move back above `a_stray_update_while_working_must_not_reinvoke_the_handler`.

**Comment by joshrotenberg** [comment]
All four addressed.

**[P1] Carry-forward vs reuse.** You are right, and the example is exact: `{approval} -> {approval, region}` was rejected. Because `requests` replaces the whole snapshot, a still-outstanding key has to be reissued to stay outstanding, so reissuing it names the same question rather than a second one. A live key is now reuse only if the request behind it changed, compared on the serialized form since `InputRequest` is `#[non_exhaustive]` and its params are not `Eq`. A carried-forward key is also no longer recorded as superseded, which was the second half of the bug: my first pass superseded the entire outgoing snapshot, so a carried key looked spent on the *next* call even if this one had allowed it.

Two tests added: retaining A while adding B, and repointing a live key at a different request (rejected).

**Error taxonomy.** Agreed that `Backend` was wrong. Added `TaskStoreError::InvalidTransition`, documented as deterministic and not retryable. The enum doc no longer claims `MemoryTaskStore` never returns errors, since it now returns this one. The router distinguishes them: a forbidden request logs at error level as a handler bug, a backend failure logs at warn as infrastructure. Both still fail the task, because a lost park leaves it unanswerable either way, and the messages say which happened.

**Contradictory contract.** The paragraph above my new section still said a reissued key is a fresh question. Rewritten, and I added the custom-store requirements you asked for: the same rule applies, and the check plus snapshot replacement must be atomic or two concurrent parks could each see a key as unspent. Also called out explicitly that stores written before this rule keep compiling and keep the old permissive behaviour, so this is a behavioural migration rather than a compile-visible one. Worth a release-note line; say the word and I will add one to the changelog on the release branch as we did for 0.21.0.

**Doc comment.** Moved back above `a_stray_update_while_working_must_not_reinvoke_the_handler`. Same scripted-insertion mistake as the orphaned rustdoc earlier in this branch series.

Full suite, clippy, `-Dmissing-docs`, and the 17-entry feature matrix are green.
